nipype/__init__.py

```diff
@@ -59,16 +59,51 @@ def get_info():
 from .interfaces import (DataGrabber, DataSink, SelectFiles, IdentityInterface,
                          Rename, Function, Select, Merge)
 
+etelemetry_results = {}
+
+def check_latest_version(raise_exception=False):
+    """Check for the latest version of the library
+
+    parameters:
+    raise_exception: boolean
+        Raise a RuntimeError if a bad version is being used
+    """
+    if raise_exception in etelemetry_results:
+        return etelemetry_results[raise_exception]
 
-if config.getboolean('execution', 'check_version'):
     import etelemetry
+    logger = logging.getLogger('nipype.utils')
 
-    latest = {"version": 'Unknown'}
+    INIT_MSG = "Running {packname} version {version} (latest: {latest})".format
+
+    latest = {"version": 'Unknown', "bad_versions": []}
+    result = None
     try:
-        latest = etelemetry.get_project("nipy/nipype")
+        result = etelemetry.get_project("nipy/nipype")
     except Exception as e:
         logger.warning("Could not check for version updates: \n%s", e)
     finally:
-        logger.info(INIT_MSG(packname='nipype',
-                             version=__version__,
-                             latest=latest["version"]))
+        if result:
+            latest.update(**result)
+            if LooseVersion(__version__) != LooseVersion(latest["version"]):
+                logger.info(INIT_MSG(packname='nipype',
+                                     version=__version__,
+                                     latest=latest["version"]))
+            if latest["bad_versions"] and \
+                    any([LooseVersion(__version__) == LooseVersion(ver)
+                         for ver in latest["bad_versions"]]):
+                message = ('You are using a version of Nipype with a critical '
+                           'bug. Please use a different version.')
+                if raise_exception:
+                    raise RuntimeError(message)
+                else:
+                    logger.critical(message)
+    etelemetry_results[raise_exception] = latest
+    return latest
+
+# Run telemetry on import for interactive sessions, such as IPython, Jupyter notebooks, Python REPL
+if config.getboolean('execution', 'check_version'):
+    import __main__
+    if not hasattr(__main__, '__file__'):
+        from .interfaces.base import BaseInterface
+        BaseInterface.check_version = check_latest_version()
```

Comment:
We're kind of doubling-up on memoization now by including it in `etelemetry_results[False]` and `BaseInterface.check_version`. What's the advantage?

Reply:
none - just lazily did not remove :)